File: app/src/main/java/com/example/getfood/data/DataManager.java
package com.example.getfood.data;

import com.example.canteen_app_models.models.CartItem;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

public class DataManager {

    private static ArrayList<CartItem> cartItems;
    private static DataManager dataManager;
    private static int cartTotal;

    public static synchronized DataManager getInstance() {
        if (dataManager == null) {
            dataManager = new DataManager();
            cartItems = new ArrayList<>();
        }
        return dataManager;
    }

    public ArrayList<CartItem> getCartItems() {
        return cartItems;
    }

    public void setCartItems(ArrayList<CartItem> cartItems) {
        DataManager.cartItems = cartItems;
    }

    public int getCartSize() {
        if (cartItems != null) {
            return cartItems.size();
        } else {
            return 0;
        }
    }

    public int getCartTotal() {
        return cartTotal;
        /*int total = 0;
        for (CartItem item : cartItems) {
            total = total + Integer.parseInt(item.getItemPrice()) * item.getItemQuantity();
        }
        return total;*/
    }

    public void increaseCartItemQuantity(int position) {
        cartItems.get(position).increaseQuantity();
        cartTotal += Integer.parseInt(cartItems.get(position).getItemPrice());
    }

    public void sortCartItems() {
        Collections.sort(cartItems, new Comparator<CartItem>() {
            @Override
            public int compare(CartItem o1, CartItem o2) {
                return o1.getItemCategory().length() - o2.getItemCategory().length();
            }
        });
    }

    public void clearCartItems() {
        cartItems.clear();
    }


    public void addItemToCart(CartItem removedItem, int position) {
        cartItems.add(position, removedItem);
        cartTotal += (removedItem.getItemQuantity() * Integer.parseInt(removedItem.getItemPrice()));
    }

    public void removeItemFromCart(int position) {
        CartItem removedItem = cartItems.remove(position);
        cartTotal -= (removedItem.getItemQuantity() * Integer.parseInt(removedItem.getItemPrice()));
    }

    public void clearCart() {
        cartItems.clear();
        cartTotal = 0;
    }

    public void decreaseCartItemQuantity(int position) {
        cartItems.get(position).decreaseQuantity();
        cartTotal -= Integer.parseInt(cartItems.get(position).getItemPrice());
    }
}

File: app/src/main/java/com/example/getfood/data/DataManager.kt
package com.example.getfood.data

import com.example.canteen_app_models.models.CartItem
import java.util.*

class DataManager {
    var cartItems: MutableList<CartItem>
        get() {
            return Companion.cartItems
        }
        set(cartItems) {
            Companion.cartItems = cartItems
        }
    val cartSize: Int
        get() {
            return Companion.cartItems.size
        }
    val cartTotal: Int
        get() {
            var total = 0
            for (item: CartItem? in Companion.cartItems) {
                total += item!!.itemPrice!!.toInt() * (item.itemQuantity)!!
            }
            return total
        }

    fun increaseCartItemQuantity(position: Int) {
        Companion.cartItems[position].increaseQuantity()
    }

    fun sortCartItems() {
        Companion.cartItems.sortWith { o1, o2 ->
            o1.itemCategory!!.length - o2.itemCategory!!.length
        }
    }

    fun clearCartItems() {
        Companion.cartItems.clear()
    }

    companion object {
        private var cartItems: MutableList<CartItem> = mutableListOf()
        private lateinit var dataManager: DataManager

        @get:Synchronized
        val instance: DataManager
            get() {
                if (!this::dataManager.isInitialized) {
                    dataManager = DataManager()
                    cartItems = ArrayList()
                }
                return dataManager
            }
    }
}